lib/http2/connection.c

```diff
@@ -82,63 +80,58 @@ static void enqueue_goaway(h2o_http2_conn_t *conn, int errnum, h2o_iovec_t addit
-    }
+    close_connection(conn);
 }
 
 static void graceful_shutdown_resend_goaway(h2o_timer_t *entry)
 {
-    h2o_context_t *ctx = H2O_STRUCT_FROM_MEMBER(h2o_context_t, http2._graceful_shutdown_timeout, entry);
-    h2o_linklist_t *node;
-    int do_close_stragglers = 0;
+    h2o_http2_conn_t *conn = H2O_STRUCT_FROM_MEMBER(h2o_http2_conn_t, _graceful_shutdown_timeout, entry);
+
+    if (conn->state < H2O_HTTP2_CONN_STATE_HALF_CLOSED) {
+        enqueue_goaway(conn, H2O_HTTP2_ERROR_NONE, (h2o_iovec_t){NULL});
 
-    for (node = ctx->http2._conns.next; node != &ctx->http2._conns; node = node->next) {
-        h2o_http2_conn_t *conn = H2O_STRUCT_FROM_MEMBER(h2o_http2_conn_t, _conns, node);
-        if (conn->state < H2O_HTTP2_CONN_STATE_HALF_CLOSED) {
-            enqueue_goaway(conn, H2O_HTTP2_ERROR_NONE, (h2o_iovec_t){NULL});
-            do_close_stragglers = 1;
+        /* After waiting a second, we still had active connections. If configured, wait one
+         * final timeout before closing the connections */
+        if (conn->super.ctx->globalconf->http2.graceful_shutdown_timeout > 0) {
+            conn->_graceful_shutdown_timeout.cb = graceful_shutdown_close_stragglers;
+            h2o_timer_link(conn->super.ctx->loop, conn->super.ctx->globalconf->http2.graceful_shutdown_timeout,
+                           &conn->_graceful_shutdown_timeout);
         }
     }
+}
 
-    /* After waiting a second, we still had active connections. If configured, wait one
-     * final timeout before closing the connections */
-    if (do_close_stragglers && ctx->globalconf->http2.graceful_shutdown_timeout > 0) {
-        ctx->http2._graceful_shutdown_timeout.cb = graceful_shutdown_close_stragglers;
-        h2o_timer_link(ctx->loop, ctx->globalconf->http2.graceful_shutdown_timeout, &ctx->http2._graceful_shutdown_timeout);
-    }
+static size_t close_idle_connection(h2o_conn_t *_conn)
+{
+    h2o_linklist_unlink(&_conn->_conns);
+    h2o_linklist_insert(&_conn->ctx->_conns.shutdown, &_conn->_conns);
+    initiate_graceful_shutdown(_conn);
+    return 1;
 }
 
-static void initiate_graceful_shutdown(h2o_context_t *ctx)
+static void initiate_graceful_shutdown(h2o_conn_t *_conn)
 {
     /* draft-16 6.8
      * A server that is attempting to gracefully shut down a connection SHOULD send an initial GOAWAY frame with the last stream
      * identifier set to 231-1 and a NO_ERROR code. This signals to the client that a shutdown is imminent and that no further
      * requests can be initiated. After waiting at least one round trip time, the server can send another GOAWAY frame with an
      * updated last stream identifier. This ensures that a connection can be cleanly shut down without losing requests.
      */
-    h2o_linklist_t *node;
 
+    h2o_http2_conn_t *conn = (void *)_conn;
     /* only doit once */
-    if (ctx->http2._graceful_shutdown_timeout.cb != NULL)
+    if (conn->_graceful_shutdown_timeout.cb != NULL)
         return;
```

Comment:
I think we can remove this `if` statement (it originates from #119 but I do not recall why it was necessary), and rather do `assert(conn->_graceful_shutdown_timeout.cb == NULL)` at the beginning of this function.

Rationale is that it is cleaner to have a contract that requires the callback to be called at most only once per each connection. If we are to provide safeguard, we can consult `ctx->shutdown_requested` in `h2o_context_request_shutdown`.